README.md
# Python Cookbook     

## About
Repository contains a cookbook used in chef with a packer.json file which can create AMIs in AWS.

## Prerequisites
- packer
- chef workstation
- AWS account

## How To
- Export your environment variables (AWS access and secret keys)
- Edit packer.json to your specification:
  - subnet_id
  - ssh_keypair_name
  - ssh_private_key_file
  - region
  - source_ami_filter
  - ssh_username
  - ami_name
  - provisioners

- cd into the cloned files and run packer.json with packer

```
packer build packer.json
```

The above will create an AMI on your AWS account which can be used to make an EC2.


CHANGELOG.md
# pythonAppChef CHANGELOG

This file is used to list changes made in each version of the pythonAppChef cookbook.

# 0.1.0

Initial release.

- change 0
- change 1



spec/unit/recipes/default_spec.rb
#
# Cookbook:: pythonAppChefAlpha
# Spec:: default
#
# Copyright:: 2020, The Authors, All Rights Reserved.

require 'spec_helper'

describe 'pythonAppChefAlpha::default' do
  context 'When all attributes are default, on Ubuntu 18.04' do
    # for a complete list of available platforms and versions see:
    # https://github.com/chefspec/fauxhai/blob/master/PLATFORMS.md
    platform 'ubuntu', '18.04'

    it 'updates the sources list' do
      expect(chef_run).to update_apt_update('update_sources')
    end

    it 'installs pip3' do
      expect(chef_run).to install_package('python3-pip')
    end

    it 'installs packer' do
      expect(chef_run).to install_package('packer')
    end

    it 'installs chef' do
      expect(chef_run).to install_package('chef')
    end

    it 'creates the Downloads folder with the correct permissions' do
      expect(chef_run).to create_directory('/home/ubuntu/Downloads').with(mode: '0007')
    end

    it 'creates the jenkins folder with the correct permissions' do
      expect(chef_run).to create_directory('/home/jenkins').with(mode: '0731')
    end

    it 'converges successfully' do
      expect { chef_run }.to_not raise_error
    end
  end
end


test/integration/default/default_test.rb
# InSpec test for recipe pythonAppChefAlpha::default

# The InSpec reference, with examples and extensive documentation, can be
# found at https://www.inspec.io/docs/reference/resources/

describe package('python3.6') do
  it { should be_installed }
end

describe package('python3-pip') do
  it { should be_installed }
end

describe package('packer') do
  it { should be_installed }
end

describe package('chef') do
  it { should be_installed }
end

describe package('openjdk-8-jdk') do
  it { should be_installed }
end

describe package('openjdk-8-jre') do
  it { should be_installed }
end

describe file('/home/ubuntu/Downloads') do
  it { should exist }
  its('mode') { should cmp '0007' }
end

describe file('/home/jenkins') do
  it { should exist }
  its('mode') { should cmp '0731' }
end

describe file('/usr/local/lib/python3.6/dist-packages/certifi') do
  it { should exist }
end

describe file('/usr/local/lib/python3.6/dist-packages/idna') do
  it { should exist }
end

describe file('/usr/local/lib/python3.6/dist-packages/six.py') do
  it { should exist }
end

describe file('/usr/local/lib/python3.6/dist-packages/urllib3') do
  it { should exist }
end

describe file('/usr/local/lib/python3.6/dist-packages/requests') do
  it { should exist }
end

describe file('/usr/local/lib/python3.6/dist-packages/atomicwrites') do
  it { should exist }
end

describe file('/usr/local/lib/python3.6/dist-packages/bs4') do
  it { should exist }
end

describe file('/usr/local/lib/python3.6/dist-packages/importlib_metadata') do
  it { should exist }
end

describe file('/usr/local/lib/python3.6/dist-packages/more_itertools') do
  it { should exist }
end

describe file('/usr/local/lib/python3.6/dist-packages/packaging') do
  it { should exist }
end

describe file('/usr/local/lib/python3.6/dist-packages/pluggy') do
  it { should exist }
end

describe file('/usr/local/lib/python3.6/dist-packages/py') do
  it { should exist }
end

describe file('/usr/local/lib/python3.6/dist-packages/pyparsing.py') do
  it { should exist }
end

describe file('/usr/local/lib/python3.6/dist-packages/pytest.py') do
  it { should exist }
end

describe file('/usr/local/lib/python3.6/dist-packages/soupsieve') do
  it { should exist }
end

describe file('/usr/local/lib/python3.6/dist-packages/wcwidth') do
  it { should exist }
end

describe file('/usr/local/lib/python3.6/dist-packages/zipp.py') do
  it { should exist }
end


recipes/default.rb
#
# Cookbook:: pythonAppChefAlpha
# Recipe:: default
#
# Copyright:: 2020, The Authors, All Rights Reserved.

apt_update 'update_sources' do
  action :update
end

package 'packer'

package 'python3-pip'

bash 'install_requirements' do
  code <<-EOH
    sudo pip3 install atomicwrites==1.3.0
    sudo pip3 install attrs==19.1.0
    sudo pip3 install beautifulsoup4==4.8.0
    sudo pip3 install certifi==2019.6.16
    sudo pip3 install chardet==3.0.4
    sudo pip3 install idna==2.8
    sudo pip3 install importlib-metadata==0.19
    sudo pip3 install more-itertools==7.2.0
    sudo pip3 install packaging==19.1
    sudo pip3 install pluggy==0.12.0
    sudo pip3 install py==1.8.0
    sudo pip3 install pyparsing==2.4.2
    sudo pip3 install pytest==5.1.0
    sudo pip3 install requests==2.22.0
    sudo pip3 install six==1.12.0
    sudo pip3 install soupsieve==1.9.2
    sudo pip3 install urllib3==1.25.3
    sudo pip3 install wcwidth==0.1.7
    sudo pip3 install zipp==0.5.2
  EOH
end

bash 'install_java' do
  code <<-EOH
    sudo apt-get install openjdk-8-jdk openjdk-8-jre -y
  EOH
end

bash 'java_env_vars' do
  code <<-EOH
    cd /etc
    sudo chmod 777 environment
    cat >> /etc/environment <<EOL
    JAVA_HOME= /usr/lib/jvm/java-8-openjdk-amd64
    JRE_HOME=/usr/lib/jvm/java-8-openjdk-amd64/jre
    EOL
    sudo chmod 644 environment
  EOH
end

bash 'install_chef' do
  code <<-EOH
    wget https://packages.chef.io/files/stable/chef-workstation/0.17.5/ubuntu/18.04/chef-workstation_0.17.5-1_amd64.deb
    sudo dpkg -i chef-workstation_*.deb
    sudo rm chef-workstation_*.deb
  EOH
end

directory '/home/ubuntu/Downloads' do
  mode '0007'
  action :create
end

directory '/home/jenkins' do
  mode '0777'
  action :create
end
